File: backend/src/main/java/com/grupocarles/admin/landsurveys/admin_landsurveys/model/LandSurvey.java
package com.grupocarles.admin.landsurveys.admin_landsurveys.model;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder

@Entity
@Table(name = "land_survey")
public class LandSurvey {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Long id;

    @Column (updatable = false)
    @CreationTimestamp
    private LocalDateTime creationDate;

    private String address;

    private Boolean corner;
    private Boolean unworkable = false;
    private String measurements;
    private BigDecimal surface;
    private LocalDateTime priceVerificationDate;
    private LocalDateTime reassessmentDate;

    @Column(columnDefinition = "TEXT")
    private String observation;

    @Column(columnDefinition = "BOOLEAN DEFAULT false")
    private Boolean isRescinded = false;

    @Column(columnDefinition = "BOOLEAN DEFAULT false")
    private Boolean isArchived = false;
    private Long price;

    //FOREIGN KEYS
    
    @ManyToOne
    @JoinColumn(name = "surveyor_id")
    private UserSec surveyor;

    @ManyToOne
    @JoinColumn(name = "manager_id")
    private UserSec manager;

    @ManyToOne
    @JoinColumn(name = "file_type")
    private FileType fileType;

    @ManyToOne
    @JoinColumn(name = "road_type")
    private RoadType roadType;

    @ManyToOne
    @JoinColumn(name = "locality_id")
    private Locality locality;

    @ManyToOne
    @JoinColumn(name = "section_id")
    private Section section;

    @ManyToOne
    @JoinColumn(name = "zone_id")
    private Zone zone;

    @ManyToOne
    @JoinColumn(name = "source")
    private Source source;

    @ManyToOne
    @JoinColumn(name = "classification")
    private Classification classification;

    @ManyToOne
    @JoinColumn(name = "agency_id")
    private Agency agency;

    @ManyToOne
    @JoinColumn(name = "particular_id", nullable = true)
    private Particular particular;

    @ManyToOne
    @JoinColumn(name = "contact_id")
    private Contact contact;

    @ManyToOne
    @JoinColumn(name = "currency_id")
    private Currency currency;

    @ManyToOne
    @JoinColumn(name = "folder_id", unique = false)
    private Folder folder;

    @ManyToOne
    @JoinColumn(name = "title_id")
    private Title title;

    @OneToMany(mappedBy = "landSurvey", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    @Builder.Default
    private List<Assessment> assessmentList = new ArrayList<>();

    public void addAssessment(Assessment assessment) {
        assessmentList.add(assessment);
        assessment.setLandSurvey(this);
    }
}

File: frontend/src/models/LandSurvey.ts
import Assessment, { emptyAssessment } from "./Assessment";
import { emptyInputOption, InputOption } from "./InputOption";

export default interface LandSurvey {
    address: string;
    agency: string;
    assessmentList?: Array<Assessment>;
    assessmentsPerSquareMeterUsd: number;
    averageAssessmentUsd: number;
    businessEvaluation: number;
    classification: string;
    contact: string;
    corner: boolean;
    unworkable: boolean;
    currency: string;
    date: string;
    daysSincePriceVerification: number;
    deviation: number;
    fileType: string;
    id: number;
    isArchived: boolean;
    isRescinded: boolean;
    locality: string;
    manager: InputOption;
    maxPrice: number;
    measurements: string;
    minPrice: number;
    observation: string;
    particular: string;
    price: number;
    pricePerSquareMeter: number;
    priceVerificationDate: string;
    reassessmentDate: string;
    roadType: string;
    section: string;
    source: string;
    surface: number;
    surveyor: InputOption;
    title: boolean;
    titleSituation: string;
    zone: string;
    folder: string;
}

export const landSurveyEmpty = (): LandSurvey => {
    return {
        address: "",
        agency: "",
        assessmentList: [emptyAssessment()],
        assessmentsPerSquareMeterUsd: 0,
        averageAssessmentUsd: 0,
        businessEvaluation: 0,
        classification: "",
        contact: "",
        corner: false,
        unworkable: false,
        currency: "",
        date: "",
        daysSincePriceVerification: 0,
        deviation: 0,
        fileType: "",
        id: 0,
        isArchived: false,
        isRescinded: false,
        locality: "",
        manager: emptyInputOption(),
        maxPrice: 0,
        minPrice: 0,
        measurements: "",
        observation: "",
        particular: "",
        price: 0,
        pricePerSquareMeter: 0,
        priceVerificationDate: "",
        reassessmentDate: "",
        roadType: "",
        section: "",
        source: "",
        surface: 0,
        surveyor: emptyInputOption(),
        title: false,
        titleSituation: "",
        zone: "",
        folder: ""
    };
};
